python/S3_VPC_ENDPOINT_ENABLED/S3_VPC_ENDPOINT_ENABLED.py

```diff
@@ -0,0 +1,389 @@
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for
+# the specific language governing permissions and limitations under the License.
+'''
+Description
+   Check whether VPC Endpoint is enabled for each VPC to access Amazon S3.
+
+Trigger
+   Periodic (because VPC Configuration Item does not record Endpoint status)
+
+Reports on:
+   AWS::EC2::VPC
+
+Rule Parameters:
+   None
+
+Scenarios:
+  Scenario 1:
+  Given: No VPC is present while using the describe_vpcs() API call
+   Then: Return NOT_APPLICABLE
+
+  Scenario 2:
+  Given: At least one VPC is present
+    And: The S3 service is not present in the list of "ServiceName" on DescribeVpcEndpoints API
+   Then: Return NON_COMPLIANT on this VPC
+
+  Scenario 3:
+  Given: At least one VPC is present
+    And: The S3 service is present in the "ServiceName" key on DescribeVpcEndpoints API
+    And: The "State" key value is not "Available"
+   Then: Return NON_COMPLIANT on this VPC
+
+  Scenario 4:
+  Given: At least one VPC is present
+    And: The S3 service is present in the "ServiceName" key on DescribeVpcEndpoints API
+    And: The "State" key value is "Available"
+   Then: Return COMPLIANT on this VPC
+'''
+import re
+import json
+import sys
+import datetime
+import boto3
+import botocore
+
+
+try:
+    import liblogging
+except ImportError:
+    pass
+
+##############
+# Parameters #
+##############
+
+# Define the default resource to report to Config Rules
+DEFAULT_RESOURCE_TYPE = 'AWS::EC2::VPC'
+
+# Set to True to get the lambda to assume the Role attached on the Config Service (useful for cross-account).
+ASSUME_ROLE_MODE = False
+
+# Other parameters (no change needed)
+CONFIG_ROLE_TIMEOUT_SECONDS = 900
+
+#############
+# Main Code #
+#############
+
+def evaluate_compliance(event, configuration_item, valid_rule_parameters):
+    vpc_list = []
+    evaluations = []
+    account_id = event['accountId']
+    ec2_client = get_client('ec2', event)
+    regex_pattern = r'com\.amazonaws\.\S*s3'
+    vpc_response = ec2_client.describe_vpcs()
+
+    try:
+        if not vpc_response['Vpcs']:
+            annotate = 'There are no Amazon VPCs in the region.'
+            evaluations.append(build_evaluation(account_id, 'NOT_APPLICABLE', event, annotation=annotate))
+        for vpc in vpc_response['Vpcs']:
+            vpc_list.append(vpc['VpcId'])
+        for vpc_id in vpc_list:
+            endpoint_response = ec2_client.describe_vpc_endpoints(Filters=[
+                {
+                    'Name':'vpc-id',
+                    'Values':[vpc_id]
+                }
+            ])
+            if not endpoint_response['VpcEndpoints']:
+                annonate = 'There are no Amazon S3 endpoints present in '+ vpc_id+'.'
+                evaluations.append(build_evaluation(vpc_id, 'NON_COMPLIANT', event, annotation=annonate))
+            for endpoint in endpoint_response['VpcEndpoints']:

```

Comment:
I found an issue here-
If a VPC has multiple endpoints configured- like S3 & EC2..this code will loop and add COMPLIANT & NON_COMPLIANT to evaluations for S3 & EC2 respectively. This cause config to report this VPC as NON_COMPLIANT. To fix this issue, you can either separate this part as a function and return from the function as soon as you get a pattern match for the VPC.

Reply:
Good catch. I changed the evaluation logic of the code, now it gives one evaluation per VPC.